ruby_programming/automated_testing/rspec_part_one_basics.md

```diff
@@ -0,0 +1,822 @@
+     # ./spec/todo_list_spec.rb:9
+
+Finished in 0.00179 seconds (files took 0.06131 seconds to load)
+3 examples, 0 failures, 3 pending
+~~~
+
+### Making Expectations
+
+Now that we have the basic structure in place, its time to write some actual test code. To test something in RSpec we make *expectations* about what the code does.
+
+To demonstrate this, let's write the test for the first `it` block. Remove the x from the `it` method to stop it from being a pending test and put the following within the block:
+
+~~~ruby
+# spec/todo_list_spec.rb
+
+it "returns the name with 0 tasks remaining" do
+  todo_list = TodoList.new("My List")
+
+  expect(todo_list.name).to eq("My List (0 tasks)")
+end
+~~~
+
+You can probably guess what this does just from reading the code, it expects the the name of the todo list we created to equal `"My List (0 tasks)"`.
+
+Run the test suite with the `$ rspec` command in your terminal. You should see that the output for the test we just filled in is now green. That means we've just wrote our first proper passing test 🎉
+
+~~~bash
+TodoList
+  name
+    returns the name with 10 tasks remaining (PENDING: Temporarily skipped with xit)
+    returns the name with 1 task remaining (PENDING: Temporarily skipped with xit)
+    returns the name with 0 tasks remaining
+
+Pending: (Failures listed here are expected and do not affect your suite's status)
+
+  1) TodoList name returns the name with 10 tasks remaining
+     # Temporarily skipped with xit
+     # ./spec/todo_list_spec.rb:13
+
+  2) TodoList name returns the name with 1 task remaining
+     # Temporarily skipped with xit
+     # ./spec/todo_list_spec.rb:10
+
+Finished in 0.00179 seconds (files took 0.06131 seconds to load)
+3 examples, 0 failures, 2 pending
+
+Randomized with seed 3657
+~~~
+
+#### Parts of an Expectation
+
+An RSpec expectation is made up of two distinct parts. The first part is the actual value we want to make the expectation on: `expect(todo_list.name).to`
+
+The second part is a matcher, which is used to verify that the actual value **matches** our expected value: `eq("My List (0 tasks)")`
+
+These two parts together make up a full expectation in RSpec:
+
+~~~ruby
+expect(actual_value).to eq(expected_value)
+~~~
+
+#### Negative Expectations
+
+Suppose you want to expect something is not to equal to something else. RSpec has you covered. You can use the `expect(actual_value).not_to` expectation, which does the opposite of `expect(actual_value).to` and will fail if the expectation **does** match.
+
+To see this in action, let's add a negative expectation to the first test case we created for the name method. In this expectation we will expect the todo list name **not to** equal `"Barrys List (0 tasks)"`
+
+~~~ruby
+# spec/todo_list_spec.rb
+
+it "returns the name with 0 tasks remaining" do
+  todo_list = TodoList.new("My List")
+
+  expect(todo_list.name).to eq("My List (0 tasks)")
+	expect(todo_list.name).not_to eq("Barrys List (0 tasks)")
```

Comment:
This line does not visually render with the tab/spaces aligned correctly. It copied correctly to VS Code, but I think in some places it is a tab & other places it is two spaces.